tools/insn_queue.c

```diff
@@ -0,0 +1,251 @@
+/*
+ *  libpulp - User-space Livepatching Library
+ *
+ *  Copyright (C) 2023 SUSE Software Solutions GmbH
+ *
+ *  This file is part of libpulp.
+ *
+ *  libpulp is free software; you can redistribute it and/or
+ *  modify it under the terms of the GNU Lesser General Public
+ *  License as published by the Free Software Foundation; either
+ *  version 2.1 of the License, or (at your option) any later version.
+ *
+ *  libpulp is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *  Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with libpulp.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "insn_queue.h"
+#include "introspection.h"
+#include "ptrace.h"
+
+#include <assert.h>
+#include <stdio.h>
+
+/** PID of target process in which we will execute the instructions.  */
+static int remote_pid;
+
+/** @brief Interpret a print instruction.
+ *
+ * @param insn    Instruction to interpet. Must be a print instruction.
+ *
+ * @return        Size of interpreted instruction.
+ */
+static int
+insn_interpret_print(struct ulp_insn *insn)
+{
+  struct ulp_insn_print *p_insn = (struct ulp_insn_print *)insn;
+
+  printf("%s\n", p_insn->bytes);
+  return insn->size;
+}
+
+/** @brief Interpret NOP instruction.
+ *
+ * @param insn    Instruction to interpet. Must be a NOP instruction.
+ *
+ * @return        Size of interpreted instruction (always 1 byte).
+ */
+static int
+insn_interpret_nop(struct ulp_insn *insn)
+{
+  (void)insn;
+  return 1;
+}
+
+/** @brief Interpret NOP instruction.
+ *
+ * @param insn    Instruction to interpet. Must be a WRITE instruction.
+ *
+ * @return        Size of interpreted instruction.
+ */
+static int
+insn_interpret_write(struct ulp_insn *insn)
+{
+  int pid = remote_pid; // Pass process pid.
+  struct ulp_insn_write *winsn = (struct ulp_insn_write *)insn;
+  if (write_bytes_ptrace(winsn->bytes, winsn->n, pid, winsn->address)) {
+    abort();
```

Comment:
The ulp tool shouldn't abort, but rather give meaningful error messages in case something goes wrong.
We might (or might not) consider leaving the target process in a defined state: e.g. if only a part of the write instructions were carried out until an error occurred `ulp` might want to backpatch whatever it wrote until then (requires reading the original content in that case of course).

Of course the backpatching itself might also not work, so it's possible that this might be overthinking it.  In any case: no aborts in tools (or libraries), only nice error handling (like you actually do in most other cases) :-)

Reply:
This went uncaught in my review, thanks!